feat: add `ndarray/base/pop`

headlessNode opened the pull request:
Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/86.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `ndarray/base/pop`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/86
-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add ndarray/base/pop

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/benchmark/benchmark.js:
```diff
@@ -0,0 +1,190 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2023 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2025 The Stdlib Authors.
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/lib/index.js:
```diff
@@ -0,0 +1,62 @@
+* arr = ndarray2array( y[ 0 ] );
+* // returns [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ]
+*
+* arr = ndarray2array( y[ 0 ] );
```
```suggestion
* arr = ndarray2array( y[ 1 ] );
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/types/index.d.ts:
```diff
@@ -0,0 +1,174 @@
+* arr = ndarray2array( y[ 0 ] );
+* // returns [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ]
+*
+* arr = ndarray2array( y[ 0 ] );
```
```suggestion
* arr = ndarray2array( y[ 1 ] );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/types/index.d.ts:
```diff
@@ -0,0 +1,174 @@
+* arr = ndarray2array( y[ 0 ] );
+* // returns [ [ 1.0, 2.0 ], [ 5.0, 6.0 ] ]
+*
+* arr = ndarray2array( y[ 0 ] );
```
```suggestion
* arr = ndarray2array( y[ 1 ] );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/types/index.d.ts:
```diff
@@ -0,0 +1,174 @@
+* arr = ndarray2array( y[ 0 ] );
+* // returns [ [ true, false ], [ true, false ] ]
+*
+* arr = ndarray2array( y[ 0 ] );
```
```suggestion
* arr = ndarray2array( y[ 1 ] );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/types/index.d.ts:
```diff
@@ -0,0 +1,174 @@
+* arr = ndarray2array( y[ 0 ] );
+* // returns [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ]
+*
+* arr = ndarray2array( y[ 0 ] );
```
```suggestion
* arr = ndarray2array( y[ 1 ] );
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/lib/main.js:
```diff
@@ -0,0 +1,87 @@
+* arr = ndarray2array( y[ 0 ] );
+* // returns [ [ 1.0, 2.0 ], [ 3.0, 4.0 ] ]
+*
+* arr = ndarray2array( y[ 0 ] );
```
```suggestion
* arr = ndarray2array( y[ 1 ] );
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/test/test.js:
```diff
@@ -0,0 +1,263 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+/* eslint-disable object-curly-newline */
```
```suggestion
```

headlessNode committed:
fix: lint error

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/repl.txt:
```diff
@@ -0,0 +1,37 @@
+    --------
+    > var x = {{alias:@stdlib/ndarray/array}}( [ [ 1, 2 ], [ 3, 4 ] ] )
+    <ndarray>
+    > var y = {{alias}}( x, 1, false, false )
+    <ndarray>
+    > {{alias:@stdlib/ndarray/to-array}}( y )
+    [ [ 2 ], [ 4 ] ]
+
```
```suggestion
    > var y = {{alias}}( x, 1, false, false )
    <ndarray>
     > {{alias:@stdlib/ndarray/to-array}}( y[0] )
    [ [ 1 ], [ 3 ] ]
    > {{alias:@stdlib/ndarray/to-array}}( y[1] )
    [ [ 2 ], [ 4 ] ]

```

headlessNode committed:
docs: fix lint error

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/repl.txt:
```diff
@@ -0,0 +1,39 @@
+    > var x = {{alias:@stdlib/ndarray/array}}( [ [ 1, 2 ], [ 3, 4 ] ] )
+    <ndarray>
+    > var y = {{alias}}( x, 1, false, false )
+    <ndarray>
```
```suggestion
    > var y = {{alias}}( x, 1, false, false )
    [ <ndarray>, <ndarray> ]
```

headlessNode committed:
docs: fix lint error

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/docs/repl.txt:
```diff
@@ -0,0 +1,39 @@
+    --------
+    > var x = {{alias:@stdlib/ndarray/array}}( [ [ 1, 2 ], [ 3, 4 ] ] )
+    <ndarray>
+    > var y = {{alias}}( x, 1, false, false )
+    [ <ndarray>, <ndarray> ]
+     > {{alias:@stdlib/ndarray/to-array}}( y[0] )
+    [ [ 1 ], [ 3 ] ]
```
```suggestion
    > var y = {{alias}}( x, 1, false, false )
    [ <ndarray>, <ndarray> ]
    > {{alias:@stdlib/ndarray/to-array}}( y[0] )
    [ [ 1 ], [ 3 ] ]
```

headlessNode committed:
docs: fix lint error

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/README.md:
```diff
@@ -0,0 +1,155 @@
+// Create a linear ndarray buffer:
+var buf = zeroTo( 27 );
+
+// Create an ndarray:
```
```suggestion
// Create an ndarray which is a stack of three 3x3 matrices:
```

Extending this comment so that the comment on L120 is a bit more obvious.

kgryte committed:
docs: update comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/README.md:
```diff
@@ -0,0 +1,155 @@
+// returns [ <ndarray>, <ndarray> ]
+
+console.log( ndarray2array( y[ 0 ] ) );
+
```
```suggestion
```

kgryte committed:
style: remove empty line

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/benchmark/benchmark.js:
```diff
@@ -0,0 +1,190 @@
+		empty( [ 2 ], { 'dtype': 'float32' }),
+		empty( [ 2 ], { 'dtype': 'int32' }),
+		empty( [ 2 ], { 'dtype': 'complex128' }),
+		empty( [ 2 ], { 'dtype': 'generic' })
```
```suggestion
		empty( [ 2 ], { 'dtype': 'float64' } ),
		empty( [ 2 ], { 'dtype': 'float32' } ),
		empty( [ 2 ], { 'dtype': 'int32' } ),
		empty( [ 2 ], { 'dtype': 'complex128' } ),
		empty( [ 2 ], { 'dtype': 'generic' } )
```

kgryte committed:
style: add missing spaces

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/lib/main.js:
```diff
@@ -0,0 +1,87 @@
+* @param {ndarray} x - input array
+* @param {integer} dim - dimension along which to perform the operation
+* @param {boolean} strict - boolean indicating whether to enforce strict bounds checking
+* @param {boolean} writable - boolean indicating whether a returned array should be writable
```
```suggestion
* @param {boolean} writable - boolean indicating whether returned arrays should be writable
```

More than one ndarray is returned.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/lib/main.js:
```diff
@@ -0,0 +1,87 @@
+	for ( i = 0; i < sh.length; i++ ) {
+		s[ i ] = null;
+	}
+	s[ dim ] = sh[ dim ] - 1;
```
You need to support negative integers for `dim`. See `ndarray/base/slice-dimension` for an example. Just means normalizing the dimension index.

headlessNode replied:
I thought I'd do normalizing in the top-level?

headlessNode replied:
For top-level signature: `pop( x[ options] )`
Made more sense to me, to validate options there.

kgryte replied:
Why should we normalize at the top-level? Why shouldn't we be able to provide a negative index to the "base" routine and have it just work?

kgryte replied:
"base" routines can validate, but they do so a bit more selectively and typically in order to avoid expensive validation steps. Normalizing indices falls into the camp of useful behavior, even for lower-level routines.

We already normalize in `slice-dimension`, `slice-dimension-from`, `slide-dimension-to`, `reverse-dimension`, `numel-dimension`, `stride`, etc, etc, etc. So it is not clear to me why `pop` should be special. Just makes it harder to remember.

headlessNode replied:
because people using "base" routine would know that `dim` can only be positive integer?

headlessNode replied:
Makes sense 👍🏻

kgryte replied:
Doing `-1` is very common in routines. Otherwise, you need to resolve an ndarray's shape and do the normalization yourself. That isn't very ergonomic is it?

kgryte replied:
Unless we have other routines in "base" which have a `dim` argument and are not normalizing, I really don't see why `pop` should be the exception. Feel free to provide such a list.

kgryte replied:
And furthermore, I just checked. For all routines which accept a `dims` argument in "base", we also normalize. https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/ndarray/base

So again, feel free to provide a stronger argument why `pop` should go against the grain here.

headlessNode replied:
No stronger argument 😄 🙌🏻

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/benchmark/benchmark.js:
```diff
@@ -0,0 +1,190 @@
+		empty( [ 2 ], { 'dtype': 'generic' } )
+	];
+
+	/* eslint-enable object-curly-newline */
```
```suggestion
	/* eslint-disable object-curly-newline, stdlib/line-closing-bracket-spacing */

	values = [
		empty( [ 2 ], { 'dtype': 'float64' } ),
		empty( [ 2 ], { 'dtype': 'float32' } ),
		empty( [ 2 ], { 'dtype': 'int32' } ),
		empty( [ 2 ], { 'dtype': 'complex128' } ),
		empty( [ 2 ], { 'dtype': 'generic' } )
	];

	/* eslint-enable object-curly-newline, stdlib/line-closing-bracket-spacing */
```

kgryte committed:
style: toggle lint rule

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/benchmark/benchmark.js:
```diff
@@ -0,0 +1,190 @@
+	var v;
+	var i;
+
+	/* eslint-disable object-curly-newline, stdlib/line-closing-bracket-spacing */
```
This applies here and below.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/README.md:
```diff
@@ -0,0 +1,154 @@
+-   **x**: input ndarray.
+-   **dim**: dimension along which to perform the operation.
+-   **strict**: boolean indicating whether to enforce strict bounds checking.
+-   **writable**: boolean indicating whether a returned ndarray should be writable.
```
```suggestion
-   **writable**: boolean indicating whether returned ndarrays should be writable.
```

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/README.md:
```diff
@@ -0,0 +1,154 @@
+The function accepts the following arguments:
+
+-   **x**: input ndarray.
+-   **dim**: dimension along which to perform the operation.
```
```suggestion
-   **dim**: dimension along which to perform the operation. If provided an integer less than zero, the dimension index is resolved relative to the last dimension, with the last dimension corresponding to the value `-1`.
```

kgryte commented on lib/node_modules/@stdlib/ndarray/base/pop/README.md:
```diff
@@ -0,0 +1,154 @@
+
+#### pop( x, dim, strict, writable )
+
+Returns an array containing a truncated view of an input ndarray and the last element along a specified dimension.
```
```suggestion
Returns an array containing a truncated view of an input ndarray and a complementary view of the last element(s) along a specified dimension.
```

Not completely happy with this reworked description, but saying "last element" might lead people to believe that they will be getting an actual element, not another ndarray view.

headlessNode committed:
refactor: apply suggestions from code review

headlessNode committed:
fix: remove wrong file

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
fix: restore file

headlessNode committed:
docs: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
docs: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
docs: apply suggestions from code review

headlessNode committed:
docs: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/lib/index.js:
```diff
@@ -0,0 +1,62 @@
+*
+* var sh = x.shape;
+* // returns [ 3, 2 ]
+*
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/ndarray/base/pop/lib/main.js:
```diff
@@ -0,0 +1,95 @@
+* var x = ndarray( 'generic', buffer, shape, strides, offset, 'row-major' );
+* // returns <ndarray>
+*
+* var sh = x.shape;
+* // returns [ 3, 2 ]
+*
```
```suggestion
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after clean-up.
